aip/general/0128.md

```diff
@@ -0,0 +1,106 @@
+
+Furthermore, there are numerous popular DevOps tools, with more being
+introduced each year. Integrating hundreds of resource types with multiple
+tools requires strong consistency, so that integration can be automated.
+
+## Guidance
+
+Services **should** clearly delineate between "control plane" operations and
+"data plane" operations, ideally through the use of distinct services with
+their own interface definition documents.
+
+- Control plane operations are responsible for managing the _lifecycle_ of
+  resources.
+- Data plane operations are responsible for managing the _content_ of
+  resources.
+
+The same resource **may** have both control plane operations and data plane
+operations associated with it. For example, a database API would have
+operations to create or delete database tables (control plane) as well as
+operations to write and read rows to that table (data plane).
+
+### Resources
+
+Resources that are declarative-friendly **must** use only strongly-consistent
+standard methods for managing resource lifecycle, which allows tools to support
+these resources generically, as well as conforming to other
+declarative-friendly guidance (see [further reading](#further-reading)).
+
+Declarative-friendly resources **should** designate that they follow the
+declarative-friendly style:
+
+```proto
+message Book {
+  option (google.api.resource) = {
+    type: "library.googleapis.com/Book"
+    pattern: "publishers/{publisher}/books/{book}"
+    style: DECLARATIVE_FRIENDLY
+  };
+
+  string name = 1;
+  // Other fields...
+}
+```
+
+### Annotations
+
+Declarative-friendly resources **must** include a
+`map<string, string> annotations` field to allow clients to store small amounts
+of arbitrary data.
+
+The `annotations` field **must** use the [Kubernetes limits][] to maintain wire
+compatibility, and **should** require dot-namespaced annotation keys to prevent
+tools from trampling over one another.
+
+**Note:** Annotations are distinct from various forms of labels. Labels can be
+used by server-side policies, such as IAM conditions. Annotations exist to
+allow client tools to store their own state information without requiring a
+database.
+
+<!-- prettier-ignore -->
+[kubernetes limits]: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set
+
+### Reconciliation
+
+If a resource takes time (more than a few seconds) for updates to be realized,
+the service **should** include an output only `bool reconciling` field to
+disclose that changes are in flight.
+
+A resource **must** set the `reconciling` field to `true` if the current state
+of the resource does not match the user's intended state, and the system is
+working to reconcile them. This is regardless of whether the root cause of
+going into reconciliation was user or system action.
+
+**Note:** Services responding to a `GET` request **must** return the resource's
```

Comment:
And we landed on there not being any way of finding out what the "currently intended state" is, right? (On the grounds that we can add that later, maybe? I can see it being useful...)

Reply:
Yes (And yes.)

TLDR: We want it. It is hard.

Reply:
I opened #699 to ask about this, before being hinted at this thread by a coworker. Would it make sense to continue the discussion there? Would you mind unofficially shedding light on the tradeoffs of various approaches at that place? I'm currently at the point where I'm considering alternatives, so I'd love to get some insights.